quickwit-cluster/docs/kubernetes-discovery.md

```diff
@@ -0,0 +1,220 @@
+- a `$release-headless` service, to be used for discovery (see next section). It will be set:
+  - as [headless](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services), to allow
+    enumerating all pods at the DNS level
+  - with `publishNotReadyAddresses: true` to expose all pods, even before they are ready
+  - with a label selector matching both the `$release-indexing` and `$release-search` pods, as a unified cluster
+
+## Proposed solution
+
+The following action items are ordered by priority, and could be independently implemented in several releases.
+Alternatives are discussed in the next section, with a pros/cons comparison.
+
+### 1a. Source the sibling pods as seeds via DNS enumeration on startup
+
+A new `peer_seed_service` configuration option is added to Quickwit. It is exclusive to the existing `peer_seeds`,
+and setting both will result on a validation error. `quickwit-cluster::start_cluster_service` is updated to check
+this option. If it is not empty:
+
+- a DNS enumeration request is issued, that will return the IP addresses of all the live pods in the cluster,
+- if the DNS request fails or if the result is empty, Quickwit fails to start. Kubernetes will restart the
+  container, allowing to retry the request when more pods are created. In case of repeated failures (misconfiguration,
+  or DNS error), the CrashLoopBackoff mechanism will engage, reducing the pressure on the DNS server by keeping
+  pods down longer between retries,
+- if not empty, the response is used to populate the `seed_nodes` argument passed to `scuttlebutt`.
+
+Due to the eventually consistent nature of the k8s control plane, there will be a delay before all pods are exposed
+in the DNS record. We do have guarantees that the responses to the different pods will be consistent. Pods will
+restart until one IP is registered, and the same first peer will be returned to all pods, avoiding a split brain.
+
+When adding nodes to the cluster (by scaling up the deployment/statefulset), new nodes will retrieve
+[up to 1000](https://kubernetes.io/docs/concepts/services-networking/service/#over-capacity-endpoints)
+IPs as seeds by querying the service. Existing nodes will not update their seeds, but discover the new nodes
+via the gossip messages.
+
+If an existing node restarts (due to an out-of-memory error for example), it will request up-to-date seeds again
+from the DNS servers, discarding its previous state. This is better than running the DNS enumeration in an
+[init container](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/), which would not
+update the seeds on container restart.
+
+### 2. Gate cluster membership on new a `cluster_name` property
+
+Redis, Cassandra and Elasticsearch all had to solve the IP reuse problem before. They all settled on the simplest and
+most resilient solution: adding a `cluster_name` property to the nodes, and forbidding nodes to join if their
```

Comment:
that sounds great.

Reply:
Using just cluster_name doesn't sound too resilient when someone misconfigures cluster_name. I agree cluster_name should be there but i would like to suggest using some kind of mTLS based communication so that a staging env node can't join production cluster.

Reply:
Thanks for the feedback @chhetripradeep.
Allowing network traffic between staging and prod is usually discouraged (as it can open a privilege escalation path to an intruder) but we cannot assume all users have effective network segmentation.

Adding mTLS on top should be doable in a subsequent release, provided the cluster operators provide the cert and key via an init-container, or a secret.